### PR TITLE
[BoundsSafety] WIP: Make 'counted_by' work for pointer fields; late parsing for 'counted_by' on decl attr position

### PR DESCRIPTION
**Pull request by rapidsna**
This contains a minimum change to make 'counted_by' available for pointer fields and make late parsing work for 'counted_by' on declaration attribute position.

**Here are TODOs:**
- As we promised in [the RFC discussion](https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854/68?u=rapidsna), separately guard the late parsing logic under a feature flag (e.g., -fexperimental-late-parse-attributes)
- Make it an error to use 'counted_by' on size-less types (e.g., forward declared structs)

**TODOs that may be done as separate PRs:**
- Fix some test cases that don't work properly for anonymous structs
- Make late parsing work for the type attribute positions
- Have __builtin_dynamic_object_size and UBSan's array-bounds checks take advantage of counted_by on pointer fields (I believe @bwendling has been working on it).
- With -fbounds-safety, make 'counted_by' and the late parsing logic available for more subjects (parameters, variables, return types, etc.)

test-edit

### CODE REVIEW

File: clang/test/Sema/attr-counted-by-late-parsed.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s
+
+#define __counted_by(f)  __attribute__((counted_by(f)))
+
+struct size_unknown;
+
+struct at_pointer {
+  int count;
+  struct size_unknown *__counted_by(count) buf; // expected-error{{'counted_by' cannot be applied to an sized type}}
```
Comment (rapidsna) [inline clang/test/Sema/attr-counted-by-late-parsed.c:9]: TODO: emit an error for 'counted_by' on unknown size like this and update the error message here.

File: clang/lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -4872,6 +4886,79 @@ void Parser::ParseStructDeclaration(
...
+  // Append the current token at the end of the new token stream so that it
+  // doesn't get lost.
+  LA.Toks.push_back(Tok);
+  PP.EnterTokenStream(LA.Toks, true, /*IsReinject=*/true);
+  // Consume the previously pushed token.
+  ConsumeAnyToken(/*ConsumeCodeCompletionTok=*/true);
+
+  ParsedAttributes Attrs(AttrFactory);
+
+  assert(
```
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:4924]: Should this be a diagnostic rather than an assert?

File: clang/lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -4872,6 +4886,79 @@ void Parser::ParseStructDeclaration(
...
+      LA.Decls.size() < 2 &&
+      "late field attribute expects to have at most a single declaration.");
+
+  Decl *D = LA.Decls.empty() ? nullptr : LA.Decls[0];
+
+  // If the Decl is on a function, add function parameters to the scope.
+  {
+    std::unique_ptr<ParseScope> Scope;
```
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:4932]: Are we missing tests for this? This looks like its to support `__counted_by()` on function parameters but I don't think we have that in tests.

File: clang/lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -4997,7 +5087,11 @@ void Parser::ParseStructUnionBody(SourceLocation RecordLoc,
 
   ParsedAttributes attrs(AttrFactory);
   // If attributes exist after struct contents, parse them.
-  MaybeParseGNUAttributes(attrs);
+  MaybeParseGNUAttributes(attrs, &LateFieldAttrs);
+
+  assert(!getLangOpts().CPlusPlus);
```
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:5092]: Should we remove the assert? Are we restricting late parsing to C only?

File: clang/lib/Parse/ParseDecl.cpp (hunk 3 above)
Comment (delcypher) [reply clang/lib/Parse/ParseDecl.cpp:4932]: Actually I guess this isn't testable because of the `  let Subjects = SubjectList<[Field], ErrorDiag>;`?

File: clang/lib/Parse/ParseDecl.cpp (hunk 4 above)
Comment (rapidsna) [reply clang/lib/Parse/ParseDecl.cpp:5092]: @delcypher C++ already has late parsing support for C++ classes and structs and I believe it's handled separately in `ParseDeclCXX.cpp`. I added the assertion to ensure that the code path isn't taken in C++. We should run clang tests if that breaks anything.

File: clang/include/clang/Driver/Options.td [hunk 5]
```diff
@@ -1599,6 +1599,13 @@ defm double_square_bracket_attributes : BoolFOption<"double-square-bracket-attri
   LangOpts<"DoubleSquareBracketAttributes">, DefaultTrue, PosFlag<SetTrue>,
  NegFlag<SetFalse>>;
 
+defm experimental_late_parse_attributes : BoolFOption<"experimental-late-parse-attributes",
```
Comment (delcypher) [inline clang/include/clang/Driver/Options.td:1602]: @rapidsna FYI. Here's a sketch of an experimental late parsing flag. It's not quite done yet and I think it's big enough that it deserves it's own PR. Just pushing it here to give you an idea of my plans

File: clang/lib/Parse/ParseDecl.cpp (hunk 4 above)
Comment (bwendling) [reply clang/lib/Parse/ParseDecl.cpp:5105]: Is it possible to combine C++'s late parsing with your implementation? Or is there little overlap?

File: clang/test/Sema/attr-counted-by-late-parsed.c (hunk 1 above)
Comment (bwendling) [reply clang/test/Sema/attr-counted-by-late-parsed.c:9]: This should be fine, because `count` is declared before use&mdash;unless I'm getting type attributes confused with field attributes.. Also, there's a grammar-o in the message: `to _an_ sized type`.

File: clang/test/Sema/attr-counted-by-late-parsed.c
Comment (bwendling) [inline clang/test/Sema/attr-counted-by-late-parsed.c:1]: FYI: I'm planning on having a separate `test/Sema/attr-counted-by-for-pointers.c` test file for pointers and rename the current file something like `test/Sema/attr-counted-by-for-flexible-array-members.c`.

Comment (rapidsna) [reply clang/test/Sema/attr-counted-by-late-parsed.c:1]: > This should be fine, because count is declared before use—unless I'm getting type attributes confused with field attributes..

`struct size_unknown *__counted_by(count) buf` indicates that `buf` has `sizeof(struct size_unknown) * count` bytes. So the problem is here that the compiler doesn't know what's the actual byte size of `buf` because `sizeof(struct unknown_size)` is not available here.

**Comment by delcypher** [comment]
@rapidsna @bwendling I've put the experimental feature flag for late parsing in a separate PR (#88596). I've not updated this PR to be based on that PR yet but will do soon. I figured the code for experimental feature flag was large enough that it deserved its own PR to make things easier to review.

**Comment by delcypher** [comment]
The implementation of the experimental late attribute changed (#88596) so I've rebased this PR on it and fixed it up to work as intended.

**Comment by delcypher** [comment]
@rapidsna I've addressed the "Make it an error to use 'counted_by' on size-less types (e.g., forward declared structs)" TODO.

I still need to test this PR properly and write a proper commit message explaining what this does.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 6]
```diff
@@ -8547,15 +8547,25 @@ static const RecordDecl *GetEnclosingNamedOrTopAnonRecord(const FieldDecl *FD) {
...
+    llvm::SmallVectorImpl<TypeCoupledDeclRefInfo> &Decls) {
   if (FD->getParent()->isUnion()) {
     S.Diag(FD->getBeginLoc(), diag::err_counted_by_attr_in_union)
         << FD->getSourceRange();
     return true;
   }
 
+  const auto FieldTy = FD->getType();
+  if (FieldTy->isPointerType() &&
+      FieldTy->getPointeeType()->isIncompleteType()) {
```
Comment (rapidsna) [inline clang/lib/Sema/SemaDeclAttr.cpp:8561]: `isIncompleteType()` won't be sufficient. We should also prevent a pointer to sizeless type, function type, and struct with flexible array member, etc.

**Comment by delcypher** [comment]
@rapidsna I've put my version of this PR in my own pull request because when I edit the PR summary here it looks like you wrote it which is weird: https://github.com/llvm/llvm-project/pull/90786

**Comment by delcypher** [comment]
@rapidsna You may want to close this PR but I'll leave this up to you.
